### PR TITLE
SWIFT-1161 MongoConnectionString authentication options support

### PR DESCRIPTION
**Pull request by bynn**
This currently doesn't have the accepted behavior for specifying the database and the authSource and I wanted to get some feedback on how I can improve the design

### CODE REVIEW

File: Tests/MongoSwiftTests/AuthTests.swift [hunk 1]
```diff
@@ -27,18 +27,18 @@ final class AuthTests: MongoSwiftTestCase {
 
         for (_, file) in testFiles {
             for testCase in file.tests {
+                print(testCase.uri, testCase.valid)
```
Comment (bynn) [inline Tests/MongoSwiftTests/AuthTests.swift:30]:
```suggestion
```

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 2]
```diff
@@ -119,6 +119,31 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
         }
     }
 
+    private static func percentDecode(from: String?) throws -> String? {
+        if from == nil {
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:123]: style nit: I think using `guard let` to unwrap `from` would be more idiomatic here:
```swift
guard let from = from else {
   return nil
}
// use non-optional `from` down here...
```

generally, when you are checking some precondition for a parameter that is necessary for the rest of the method to execute, `guard` is appropriate

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 3]
```diff
@@ -119,6 +119,31 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
...
+            return nil
+        }
+        let decodedOptional = from?.removingPercentEncoding
+        guard let decoded = decodedOptional else {
+            throw MongoError.InvalidArgumentError(
+                message: "Contains an unescaped percent sign"
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:129]:
```suggestion
                message: "Connection string contains an unescaped percent sign"
```

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 4]
```diff
@@ -119,6 +119,31 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
         }
     }
 
+    private static func percentDecode(from: String?) throws -> String? {
+        if from == nil {
+            return nil
+        }
+        let decodedOptional = from?.removingPercentEncoding
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:126]: anther nit - you can one-line this, too, and save a variable:
`guard let decoded = from.removingPercentEncoding else {`

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 5]
```diff
@@ -189,4 +358,9 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
 
     /// Specifies one or more host/ports to connect to.
     public var hosts: [HostIdentifier]
+
+    /// Returns the credential configured on this URI. Will be empty if no options are set.
+    public var credential: MongoCredential?
+
+    public var database: String?
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:365]: this property needs a docstring

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 6]
```diff
@@ -156,8 +181,109 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
...
+                guard keyValue.count == 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "Invalid key value pair for option \(keyValue)"
+                    )
+                }
+                let key = keyValue[0]
+                let value = keyValue[1]
+                switch key {
+                // authentication options.
+                case "authSource":
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:220]: related to my comment above, perhaps you could break out parsing auth options into a separate helper method, and figure out a way to consolidate the repeated check for whether `self.credential != nil`. 

I think what you're doing here to avoid multiple linear scans through the array is a good idea, so maybe you could continue supporting that by e.g. parsing out options into a `[String: String]` that you pass around into helpers that parse specific options?

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 7]
```diff
@@ -180,7 +306,50 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
...
         des += "\(self.scheme)://"
+        if let username = self.credential?.username {
+            des += username
+            if let password = self.credential?.password {
+                des += ":\(password)"
+            }
+            des += "@"
+        }
         des += self.hosts.map { $0.description }.joined(separator: ",")
+//        if let database = self.database {
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:317]: should this be uncommented?

**Review by kmahar — CHANGES_REQUESTED** [review]

File: Sources/MongoSwift/MongoConnectionString.swift (hunk 6 above)
Comment (bynn) [reply Sources/MongoSwift/MongoConnectionString.swift]: authOptions is just a temporary variable but is accessible if needed 
options will not be applied and verified in separate functions

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 8]
```diff
@@ -156,8 +183,115 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
                     message: "A port cannot be specified in a mongodb+srv connection string")
             }
         }
+        let authAndOptions = identifiersAndOptions.count == 2 ?
```
[the remark this answers is not in the document]
Comment (bynn) [reply Sources/MongoSwift/MongoConnectionString.swift:235]: applyAuthOptions created

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 9]
```diff
@@ -119,6 +119,82 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
...
+                    guard propertyKeyValue.count == 2 else {
+                        throw MongoError.InvalidArgumentError(
+                            message: "Invalid key value pair for authMechanismProperties \(propertyKeyValue)"
+                        )
+                    }
+                    let propertyKey = propertyKeyValue[0]
+                    let propertyValue = propertyKeyValue[1]
+                    switch propertyKey {
+                    case "SERVICE_NAME", "SERVICE_REALM":
+                        if let cred = self.credential?.mechanism {
```
Comment (kmahar) [inline Sources/MongoSwift/MongoConnectionString.swift:157]: what if we process the auth mechanism properties before we process the mechanism option? could we run into an issue?

**Review by kmahar — CHANGES_REQUESTED** [review]
thanks for making those changes! my next batch of comments is mainly around trying to make the logic flow a little more clear for future readers of the code, and ensuring we're handling edge cases

File: Sources/MongoSwift/MongoConnectionString.swift (hunk 5 above)
Comment (bynn) [reply Sources/MongoSwift/MongoConnectionString.swift]: done

File: Tests/MongoSwiftTests/AuthTests.swift [hunk 10]
```diff
@@ -21,37 +21,65 @@ struct AuthTestCase: Decodable {
     let credential: MongoCredential?
 }
 
+extension MongoCredential {
+    func matches(testCredential: MongoCredential, description: String) throws {
+        var connStringCredential = self
```
[the remark this answers is not in the document]
Comment (bynn) [reply Tests/MongoSwiftTests/AuthTests.swift]: done

File: Tests/MongoSwiftTests/AuthTests.swift [hunk 11]
```diff
@@ -21,37 +21,65 @@ struct AuthTestCase: Decodable {
...
+        var credential = testCredential
+
+        expect(connStringCredential).toNot(beNil(), description: description)
+
+        // Ignore default properties for now.
```
[the remark this answers is not in the document]
Comment (bynn) [reply Tests/MongoSwiftTests/AuthTests.swift]: done

File: Sources/MongoSwift/MongoConnectionString.swift [hunk 12]
```diff
@@ -180,7 +308,45 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
...
+            var properties: [String] = []
+            // TODO: SWIFT-1177 Remove after BSON conforms to CustomStringConvertible
+            for (k, v) in mechanismProperty {
+                if let value = v.stringValue {
+                    properties.append("\(k):\(value)")
+                } else if let value = v.boolValue {
+                    properties.append("\(k):\(String(value))")
+                }
+            }
+            options.append("authMechanismProperties=\(properties.joined(separator: ","))")
```
[the remark this answers is not in the document]
Comment (bynn) [reply Sources/MongoSwift/MongoConnectionString.swift:374]: now checks for empty list

**Comment by kmahar** [comment]
we're closing this PR now as it now longer merges cleanly, however we will keep this branch around for when we have time to revisit the project, so we can use Bynn's work.
